File: src/main/java/com/github/srilakshmikanthanp/quicknote/utility/Preference.java
// Copyright (c) 2021 Sri Lakshmi Kanthan P
// 
// This software is released under the MIT License.
// https://opensource.org/licenses/MIT

package com.github.srilakshmikanthanp.quicknote.utility;

import java.util.prefs.Preferences;
import java.util.prefs.PreferenceChangeListener;

/**
 * User Preference class
 */
public class Preference {
    /**
     * Preference node name
     */
    private final static Preferences prefs = Preferences.userNodeForPackage(Preference.class);

    /*********************************************************/
    /**                         Height                      **/
    /*********************************************************/

    public static final String HEIGHT_PREF_KEY = "HEIGHT";
    public static final double MIN_HEIGHT = 100;
    public static final double MAX_HEIGHT = 500;
    public static final double DEFAULT_HEIGHT = 350;

    /**
     * setter of the height of the Note.
     */
    public static void setHeight(double height) {
        Preference.prefs.putDouble(HEIGHT_PREF_KEY, height);
    }
    
    /**
     * getter of the width of the Note.
     */
    public static double getHeight() {
        return Preference.prefs.getDouble(HEIGHT_PREF_KEY, DEFAULT_HEIGHT);
    }

    /********************************************************/
    /**                         Width                      **/
    /********************************************************/

    public static final String WIDTH_PREF_KEY = "WIDTH";
    public static final double MIN_WIDTH = 100;
    public static final double MAX_WIDTH = 550;
    public static final double DEFAULT_WIDTH = 350;

    /**
     * setter of the width of the Note.
     */
    public static void setWidth(double width) {
        Preference.prefs.putDouble(WIDTH_PREF_KEY, width);
    }
    
    /**
     * getter of the width of the Note.
     */
    public static double getWidth() {
        return Preference.prefs.getDouble(WIDTH_PREF_KEY, DEFAULT_WIDTH);
    }

    /********************************************************/
    /**                         Themes                     **/
    /********************************************************/

    public static final String THEME_PREF_KEY = "THEME";
    public static final String LIGHT_THEME = "Light";
    public static final String DARK_THEME = "Dark";
    public static final String DEFAULT_THEME = DARK_THEME;


    /**
     * setter of the theme of the Note.
     */
    public static void setTheme(String theme) {
        Preference.prefs.put(THEME_PREF_KEY, theme);
    }

    /**
     * getter of the theme of the Note.
     */
    public static String getTheme() {
        return Preference.prefs.get(THEME_PREF_KEY, DEFAULT_THEME);
    }

    /********************************************************/
    /**                         Lock Focus                 **/
    /********************************************************/
    
    public static final String LOCK_PREF_KEY = "LOCK_FOCUS";
    public static final boolean DEFAULT_LOCK = false;

    /**
     * setter of the lock focus of the Note.
     */
    public static void setLock(boolean lockFocus) {
        Preference.prefs.putBoolean(LOCK_PREF_KEY, lockFocus);
    }

    /**
     * getter of the lock focus of the Note.
     */
    public static boolean getLock() {
        return Preference.prefs.getBoolean(LOCK_PREF_KEY, DEFAULT_LOCK);
    }

    /********************************************************/
    /**                         Text                       **/
    /********************************************************/

    public static final String TEXT_PREF_KEY = "TEXT";
    public static final String DEFAULT_TEXT = "";

    /**
     * setter of the text of the Note.
     */
    public static void setText(String text) {
        Preference.prefs.put(TEXT_PREF_KEY, text);
    }

    /**
     * getter of the text of the Note.
     */
    public static String getText() {
        return Preference.prefs.get(TEXT_PREF_KEY, DEFAULT_TEXT);
    }

    /**
     * Preference Change Listener
     * 
     * @param listener PreferenceChangeListener
     */
    public void addPreferenceChangeListener(PreferenceChangeListener listener) {
        Preference.prefs.addPreferenceChangeListener(listener);
    }
}

File: src/main/kotlin/com/github/srilakshmikanthanp/quicknote/utility/Preference.kt
package com.github.srilakshmikanthanp.quicknote.utility

import com.github.srilakshmikanthanp.quicknote.consts.AppConsts
import java.util.prefs.Preferences
import java.util.prefs.PreferenceChangeListener

/**
 * Object that helps to store the App Preference
 */
object Preference {
    /**
     * Preference Engine from Standard Java API
     */
    private val prefEngine: Preferences = Preferences.userNodeForPackage(Preference.javaClass)

    /*******************************
     *   Width of the Editor       *
     ******************************/
    const val WIDTH_KEY = "WIDTH"

    /**
     * Width Getter from preference
     *
     * @return Height
     */
    fun getWidth() = prefEngine.getDouble(WIDTH_KEY, AppConsts.DEF_HEIGHT)

    /**
     * Width Setter to preference
     *
     * @param width
     */
    fun setWidth(width: Double) = prefEngine.putDouble(WIDTH_KEY, width)

    /*******************************
     *   Height of the Editor      *
     ******************************/
    const val HEIGHT_KEY = "HEIGHT"

    /**
     * Height Getter from preference
     *
     * @return Height
     */
    fun getHeight() = prefEngine.getDouble(HEIGHT_KEY, AppConsts.DEF_HEIGHT)

    /**
     * Height Setter to preference
     *
     * @param height
     */
    fun setHeight(height:Double) = prefEngine.putDouble(HEIGHT_KEY, height)

    /******************************
     *  Dark MODE of the Editor   *
     *****************************/
    const val DARK_KEY = "DARK_THEME"

    /**
     * Theme getter from preference
     *
     * @return true if dark else false
     */
    fun isDark() = prefEngine.getBoolean(DARK_KEY, AppConsts.DEF_THEME)

    /**
     * Theme setter to Preference
     *
     * @param isDark
     */
    fun setDark(isDark:Boolean) = prefEngine.putBoolean(DARK_KEY, isDark)

    /******************************
     *  Lock of the Editor        *
     *****************************/
    const val LOCK_KEY = "LOCKED"

    /**
     * Lock getter from preference
     *
     * @return lock
     */
    fun isLocked() = prefEngine.getBoolean(LOCK_KEY, AppConsts.DEF_LOCK)

    /**
     * Lock setter to preference
     *
     * @param isLocked
     */
    fun setLocked(isLocked:Boolean) = prefEngine.putBoolean(LOCK_KEY, isLocked)

    /*****************************
     * Text of the Editor        *
     ****************************/
    const val TEXT_KEY = "TEXT_VAL"

    /**
     * Text getter from preference
     *
     * @return text
     */
    fun getText() = prefEngine.get(TEXT_KEY, AppConsts.DEF_TEXT)

    /**
     * Text setter to preference
     *
     * @param text
     */
    fun setText(text:String) = prefEngine.put(TEXT_KEY, text)

    /***************************
     * Preference Listener     *
     **************************/

    /**
     * Add a preference Change Listener
     *
     * @param listener
     */
    fun addPreferenceListener(listener: PreferenceChangeListener) {
        prefEngine.addPreferenceChangeListener(listener)
    }
}
